server/src/main/java/org/apache/kafka/server/ClientQuotaManager.java

```diff
@@ -0,0 +1,883 @@
+    }
+
+    /**
+     * Records that a user/clientId changed some metric being throttled without checking for
+     * quota violation. The aggregate value will subsequently be used for throttling when the
+     * next request is processed.
+     */
+    public void recordNoThrottle(Session session, String clientId, double value) {
+        ClientSensors clientSensors = getOrCreateQuotaSensors(session, clientId);
+        clientSensors.quotaSensor().record(value, time.milliseconds(), false);
+    }
+
+    /**
+     * "Unrecord" the given value that has already been recorded for the given user/client by recording a negative value
+     * of the same quantity.
+     * For a throttled fetch, the broker should return an empty response and thus should not record the value. Ideally,
+     * we would like to compute the throttle time before actually recording the value, but the current Sensor code
+     * couples value recording and quota checking very tightly. As a workaround, we will unrecord the value for the fetch
+     * in case of throttling. Rate keeps the sum of values that fall in each time window, so this should bring the
+     * overall sum back to the previous value.
+     */
+    public void unrecordQuotaSensor(Session session, String clientId, double value, long timeMs) {
+        ClientSensors clientSensors = getOrCreateQuotaSensors(session, clientId);
+        clientSensors.quotaSensor().record(value * -1, timeMs, false);
+    }
+
+    /**
+     * Returns the maximum value that could be recorded without guaranteed throttling.
+     * Recording any larger value will always be throttled, even if no other values were recorded in the quota window.
+     * This is used for deciding the maximum bytes that can be fetched at once
+     */
+    public double getMaxValueInQuotaWindow(Session session, String clientId) {
+        if (quotasEnabled()) {
+            ClientSensors clientSensors = getOrCreateQuotaSensors(session, clientId);
+            Double limit = quotaCallback.quotaLimit(clientQuotaType, clientSensors.metricTags());
+            if (limit != null) {
+                return limit * (config.numQuotaSamples - 1) * config.quotaWindowSizeSeconds;
+            } else {
+                return Double.MAX_VALUE;
+            }
+        } else {
+            return Double.MAX_VALUE;
+        }
+    }
+
+    /**
+     * Throttle a client by muting the associated channel for the given throttle time.
+     * @param clientId request client id
+     * @param session request session
+     * @param throttleTimeMs Duration in milliseconds for which the channel is to be muted.
+     * @param throttleCallback Callback for channel throttling
+     */
+    public void throttle(
+            String clientId,
+            Session session,
+            ThrottleCallback throttleCallback,
+            int throttleTimeMs
+    ) {
+        if (throttleTimeMs > 0) {
+            ClientSensors clientSensors = getOrCreateQuotaSensors(session, clientId);
```

Comment:
it would be cool if we can leverage `var` to cleanup the type declaration.
```java
            var clientSensors = getOrCreateQuotaSensors(session, clientId);
            clientSensors.throttleTimeSensor().record(throttleTimeMs);
            var throttledChannel = new ThrottledChannel(time, throttleTimeMs, throttleCallback);
```